File: Try_GameEngine5/src/com/example/try_gameengine/avg/MessageBoxStringView.java
package com.example.try_gameengine.avg;

import android.app.Activity;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.PixelFormat;
import android.view.SurfaceHolder;
import android.view.SurfaceHolder.Callback;
import android.view.MotionEvent;
import android.view.SurfaceView;
import android.view.View;
import android.view.ViewGroup;
import android.widget.RelativeLayout;

public class MessageBoxStringView {
	SurfaceHolder holder;
	protected boolean visible = true;
	private char[] showMessages;
	
	private int size, tmp_left, left, fontSize, fontHeight;
	private int width, height, leftOffset, topOffset, next, messageCount;
	private int messageLength = 10;
	private int interceptMaxString;

	private int interceptCount;

	private LColor fontColor = LColor.white;
	private boolean onComplete, newLine;
	private Bitmap creeseIcon;
	private int iconWidth;
	private StringBuffer messageBuffer = new StringBuffer(messageLength);
	
	private String messages;
	
	View view;
	int x, y;
	Context context;
	LMessage message;
	
	public MessageBoxStringView(Context context) {
//		super(context);
		// TODO Auto-generated constructor stub
		
//		holder = getHolder();
//		holder.addCallback(this);
	}
	
	public MessageBoxStringView(Context context, LMessage message, int x, int y, int width, int height) {
//		super(context);
		// TODO Auto-generated constructor stub
		
//		holder = getHolder();
//		holder.addCallback(this);
		this.context = context;
		this.message = message;
		this.x = x;
		this.y = y;
		this.width = width;
		this.height = height;
	}
	
	public MessageView getView(){
		MessageView messageView = new MessageView(context, x, y, width, height);

		return messageView;
	}
	
	private void draw(GraphicsObject g) {
		draw(g, LColor.white);
	}
	
	public void draw(GraphicsObject g, LColor old) {
		if (!visible) {
			return;
		}
		float alpha = g.getAlpha();
		g.setAlpha(1.0f);
		drawMessage(g, old);
		g.setAlpha(alpha);
	}
	
	private void drawMessage(GraphicsObject g, LColor old) {
		if (!visible) {
			return;
		}
		synchronized (showMessages) {
			this.size = showMessages.length;
			this.fontSize = g.getFont().getSize();
			this.fontHeight = g.getFont().getHeight();
			this.tmp_left = (width - (fontSize * messageLength)) / 2
					- (int) (fontSize * 1.5);
			this.left = tmp_left;
			int index = 0, offset = 0, font = 0, tmp_font = 0;
			int fontSizeDouble = fontSize * 2;
			for (int i = 0; i < size; i++) {
				if (interceptCount < interceptMaxString) {
					interceptCount++;
					g.setColor(fontColor);
					continue;
				} else {
					interceptMaxString = 0;
					interceptCount = 0;
				}
				if (showMessages[i] == 'n'
						&& showMessages[i > 0 ? i - 1 : 0] == '\\') {
					index = 0;
					left = tmp_left;
					offset++;
					continue;
				} else if (showMessages[i] == '\n') {
					index = 0;
					left = tmp_left;
					offset++;
					continue;
				} else if (showMessages[i] == '<') {
					LColor color = getColor(showMessages[i < size - 1 ? i + 1
							: i]);
					if (color != null) {
						interceptMaxString = 1;
						fontColor = color;
					}
					next();
					continue;
				} else if (showMessages[i > 0 ? i - 1 : i] == '<'
						&& getColor(showMessages[i]) != null) {
					continue;
				} else if (showMessages[i] == '/') {
					if (showMessages[i < size - 1 ? i + 1 : i] == '>') {
						interceptMaxString = 1;
						fontColor = old;
					}
					continue;
				} else if (index > messageLength) {
					index = 0;
					left = tmp_left;
					offset++;
					newLine = false;
				} else if (showMessages[i] == '\\') {
					continue;
				}
				String mes = String.valueOf(showMessages[i]);
				tmp_font = g.getFont().charWidth(showMessages[i]);
				if (Character.isLetter(showMessages[i])) {
					font = tmp_font;
				} else {
					font = fontSize;
				}
				left += font;
				if (i != size - 1) {
					g.drawString(mes, x + left + leftOffset,
							(offset * fontHeight) + y + fontSizeDouble
									+ topOffset);
				} else if (!newLine && !onComplete) {
					g.drawImage(creeseIcon, x + left + leftOffset
							+ iconWidth, (offset * fontHeight) + y
							+ fontSize + topOffset);
				}
				index++;
			}
			if (messageCount == next) {
				onComplete = true;
			}
			
			g.drawImage(message.background, x, y);
		}
	}
	
	public boolean next() {
		synchronized (showMessages) {
			if (!onComplete) {
				if (messageCount == next) {
					onComplete = true;
					return false;
				}
				if (messageBuffer.length() > 0) {
					messageBuffer.delete(messageBuffer.length() - 1,
							messageBuffer.length());
				}
				this.messageBuffer.append(messages.charAt(messageCount));
				this.messageBuffer.append("_");
				this.showMessages = messageBuffer.toString().toCharArray();
				this.size = showMessages.length;
				this.messageCount++;
			} else {
				return false;
			}
			return true;
		}
	}
	
	private LColor getColor(char flagName) {
		if ('r' == flagName || 'R' == flagName) {
			return LColor.red;
		}
		if ('b' == flagName || 'B' == flagName) {
			return LColor.black;
		}
		if ('l' == flagName || 'L' == flagName) {
			return LColor.blue;
		}
		if ('g' == flagName || 'G' == flagName) {
			return LColor.green;
		}
		if ('o' == flagName || 'O' == flagName) {
			return LColor.orange;
		}
		if ('y' == flagName || 'Y' == flagName) {
			return LColor.yellow;
		} else {
			return null;
		}
	}
	
	public void setMessage(String context) {
		setMessage(context, false);
	}

	public void setMessage(String context, boolean isComplete) {
		this.visible = false;
		this.showMessages = new char[] { '\0' };
		this.interceptMaxString = 0;
		this.next = 0;
		this.messageCount = 0;
		this.interceptCount = 0;
		this.size = 0;
		this.tmp_left = 0;
		this.left = 0;
		this.fontSize = 0;
		this.fontHeight = 0;
		this.messages = context;
		this.next = context.length();
		this.onComplete = false;
		this.newLine = false;
		this.messageCount = 0;
		this.messageBuffer.delete(0, messageBuffer.length());
//		if (isComplete) {
//			this.complete();
//		}
		this.visible = true;
	}
	
	public boolean isComplete() {
		return onComplete;
	}
	
	public Bitmap getCreeseIcon() {
		return creeseIcon;
	}

	public void setCreeseIcon(Bitmap creeseIcon) {
		this.creeseIcon = creeseIcon;
		this.iconWidth = creeseIcon.getWidth();
	}

	public int getMessageLength() {
		return messageLength;
	}

	public void setMessageLength(int messageLength) {
		this.messageLength = messageLength;
	}

//	public int getHeight() {
//		return height;
//	}

	public void setHeight(int height) {
		this.height = height;
	}

//	public int getWidth() {
//		return width;
//	}

	public void setWidth(int width) {
		this.width = width;
	}

	public int getLeftOffset() {
		return leftOffset;
	}

	public void setLeftOffset(int leftOffset) {
		this.leftOffset = leftOffset;
	}

	public int getTopOffset() {
		return topOffset;
	}

	public void setTopOffset(int topOffset) {
		this.topOffset = topOffset;
	}

	public boolean isVisible() {
		return visible;
	}

	public void setVisible(boolean visible) {
		this.visible = visible;
	}
	
	public class MessageView extends SurfaceView implements Callback{
		private Canvas canvas;
		boolean isUseSelf = true;
		
		public MessageView(Context context, int x, int y, int width, int height) {
			super(context);
			holder = getHolder();
			holder.addCallback(this);
			
			holder.setFormat(PixelFormat.TRANSPARENT);
			setZOrderOnTop(true);
			setZOrderMediaOverlay(true);
		}
		
		public void setIsUseSelf(boolean isUseSelf){
			this.isUseSelf = false;
		}
		
		public void setCanvas(Canvas canvas){
			this.canvas = canvas;
		}
		
		public void draw(){
			if(isUseSelf){
				canvas = holder.lockCanvas();
				canvas.drawColor(Color.WHITE);
			}
//			canvas.drawText("123" +
//					"123", x, y, paint);
			Paint paint = new Paint();
			paint.setTextSize(message.getMessageFont().getSize());
			GraphicsObject g = new GraphicsObject(canvas, paint);
			
			message.createCustomUI(g, x, y);
//			drawMessage(g, new LColor(100, 100, 100));
			
			AVGUtils.pause(30);
			
			message.update(30);
			
			if(isUseSelf)
				holder.unlockCanvasAndPost(canvas);
		}
				
		Thread thread = new Thread(new Runnable() {
			
			@Override
			public void run() {
				// TODO Auto-generated method stub
				while(true){
					draw();
					
				}
				
			}
		});

		@Override
		public void surfaceChanged(SurfaceHolder holder, int format, int width,
				int height) {
			// TODO Auto-generated method stub
			
		}

		@Override
		public void surfaceCreated(SurfaceHolder holder) {
			// TODO Auto-generated method stub
			if(isUseSelf)
				thread.start();
			ViewGroup.LayoutParams lp = getLayoutParams();
			lp.width = width;
			lp.height = height;
			setLayoutParams(lp);
		}

		@Override
		public void surfaceDestroyed(SurfaceHolder holder) {
			// TODO Auto-generated method stub
			
		}
		
		
		
		@Override
		public boolean onTouchEvent(MotionEvent event) {
			// TODO Auto-generated method stub

//			bb(event);

			return true;
		}
	}
	

	
}

File: Try_GameEngine5/src/com/example/try_gameengine/avg/MessageBoxStringView.kt
package com.example.try_gameengine.avg

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.PixelFormat
import android.view.MotionEvent
import android.view.SurfaceHolder
import android.view.SurfaceView
import android.view.View

class MessageBoxStringView {
    var holder: SurfaceHolder? = null
    var isVisible: Boolean = true
    private var showMessages: CharArray = charArrayOf('\u0000')

    private var size = 0
    private var tmp_left = 0
    private var left = 0
    private var fontSize = 0
    private var fontHeight = 0
    private var width = 0
    private var height = 0
    var leftOffset: Int = 0
    var topOffset: Int = 0
    private var next = 0
    private var messageCount = 0
    var messageLength: Int = 10
    private var interceptMaxString = 0

    private var interceptCount = 0

    private var fontColor: LColor? = LColor.Companion.white
    var isComplete: Boolean = false
        private set
    private var newLine = false
    private var creeseIcon: Bitmap? = null
    private var iconWidth = 0
    private val messageBuffer = StringBuffer(messageLength)

    private var messages: String? = null

    var view: View? = null
    var x: Int = 0
    var y: Int = 0
    var context: Context? = null
    var message: LMessage? = null

    constructor(context: Context?)

    constructor(context: Context?, message: LMessage, x: Int, y: Int, width: Int, height: Int) {
//		super(context);
        // TODO Auto-generated constructor stub

//		holder = getHolder();
//		holder.addCallback(this);

        this.context = context
        this.message = message
        this.x = x
        this.y = y
        this.width = width
        this.height = height
    }

    fun getView(): MessageView {
        val messageView = MessageView(context, x, y, width, height)

        return messageView
    }

    private fun draw(g: GraphicsObject) {
        draw(g, LColor.Companion.white)
    }

    fun draw(g: GraphicsObject, old: LColor?) {
        if (!this.isVisible) {
            return
        }
        val alpha = g.getAlpha()
        g.alpha = 1.0f
        drawMessage(g, old)
        g.alpha = alpha
    }

    private fun drawMessage(g: GraphicsObject, old: LColor?) {
        if (!this.isVisible) {
            return
        }
        synchronized(showMessages) {
            this.size = showMessages.size
            this.fontSize = g.getFont()!!.getSize()
            this.fontHeight = g.getFont()!!.getHeight()
            this.tmp_left = ((width - (fontSize * messageLength)) / 2
                    - (fontSize * 1.5).toInt())
            this.left = tmp_left
            var index = 0
            var offset = 0
            var font = 0
            var tmp_font = 0
            val fontSizeDouble = fontSize * 2
            for (i in 0..<size) {
                if (interceptCount < interceptMaxString) {
                    interceptCount++
                    g.setColor(fontColor!!)
                    continue
                } else {
                    interceptMaxString = 0
                    interceptCount = 0
                }
                if (showMessages[i] == 'n'
                    && showMessages[if (i > 0) i - 1 else 0] == '\\'
                ) {
                    index = 0
                    left = tmp_left
                    offset++
                    continue
                } else if (showMessages[i] == '\n') {
                    index = 0
                    left = tmp_left
                    offset++
                    continue
                } else if (showMessages[i] == '<') {
                    val color = getColor(
                        showMessages[if (i < size - 1)
                            i + 1
                        else
                            i]
                    )
                    if (color != null) {
                        interceptMaxString = 1
                        fontColor = color
                    }
                    next()
                    continue
                } else if (showMessages[if (i > 0) i - 1 else i] == '<'
                    && getColor(showMessages[i]) != null
                ) {
                    continue
                } else if (showMessages[i] == '/') {
                    if (showMessages[if (i < size - 1) i + 1 else i] == '>') {
                        interceptMaxString = 1
                        fontColor = old
                    }
                    continue
                } else if (index > messageLength) {
                    index = 0
                    left = tmp_left
                    offset++
                    newLine = false
                } else if (showMessages[i] == '\\') {
                    continue
                }
                val mes = showMessages[i].toString()
                tmp_font = g.getFont()!!.charWidth(showMessages[i])
                if (Character.isLetter(showMessages[i])) {
                    font = tmp_font
                } else {
                    font = fontSize
                }
                left += font
                if (i != size - 1) {
                    g.drawString(
                        mes, x + left + leftOffset,
                        ((offset * fontHeight) + y + fontSizeDouble
                                + topOffset)
                    )
                } else if (!newLine && !this.isComplete) {
                    g.drawImage(
                        creeseIcon, (x + left + leftOffset
                                + iconWidth), ((offset * fontHeight) + y
                                + fontSize + topOffset)
                    )
                }
                index++
            }
            if (messageCount == next) {
                this.isComplete = true
            }
            g.drawImage(message!!.background, x, y)
        }
    }

    fun next(): Boolean {
        synchronized(showMessages) {
            if (!this.isComplete) {
                if (messageCount == next) {
                    this.isComplete = true
                    return false
                }
                if (messageBuffer.length > 0) {
                    messageBuffer.delete(
                        messageBuffer.length - 1,
                        messageBuffer.length
                    )
                }
                this.messageBuffer.append(messages!!.get(messageCount))
                this.messageBuffer.append("_")
                this.showMessages = messageBuffer.toString().toCharArray()
                this.size = showMessages.size
                this.messageCount++
            } else {
                return false
            }
            return true
        }
    }

    private fun getColor(flagName: Char): LColor? {
        if ('r' == flagName || 'R' == flagName) {
            return LColor.Companion.red
        }
        if ('b' == flagName || 'B' == flagName) {
            return LColor.Companion.black
        }
        if ('l' == flagName || 'L' == flagName) {
            return LColor.Companion.blue
        }
        if ('g' == flagName || 'G' == flagName) {
            return LColor.Companion.green
        }
        if ('o' == flagName || 'O' == flagName) {
            return LColor.Companion.orange
        }
        if ('y' == flagName || 'Y' == flagName) {
            return LColor.Companion.yellow
        } else {
            return null
        }
    }

    fun setMessage(context: String) {
        setMessage(context, false)
    }

    fun setMessage(context: String, isComplete: Boolean) {
        this.isVisible = false
        this.showMessages = charArrayOf('\u0000')
        this.interceptMaxString = 0
        this.next = 0
        this.messageCount = 0
        this.interceptCount = 0
        this.size = 0
        this.tmp_left = 0
        this.left = 0
        this.fontSize = 0
        this.fontHeight = 0
        this.messages = context
        this.next = context.length
        this.isComplete = false
        this.newLine = false
        this.messageCount = 0
        this.messageBuffer.delete(0, messageBuffer.length)
        //		if (isComplete) {
//			this.complete();
//		}
        this.isVisible = true
    }

    fun getCreeseIcon(): Bitmap? {
        return creeseIcon
    }

    fun setCreeseIcon(creeseIcon: Bitmap) {
        this.creeseIcon = creeseIcon
        this.iconWidth = creeseIcon.getWidth()
    }

    //	public int getHeight() {
    //		return height;
    //	}
    fun setHeight(height: Int) {
        this.height = height
    }

    //	public int getWidth() {
    //		return width;
    //	}
    fun setWidth(width: Int) {
        this.width = width
    }

    inner class MessageView(context: Context?, x: Int, y: Int, width: Int, height: Int) :
        SurfaceView(context), SurfaceHolder.Callback {
        private var canvas: Canvas? = null
        var isUseSelf: Boolean = true

        fun setIsUseSelf(isUseSelf: Boolean) {
            this.isUseSelf = false
        }

        fun setCanvas(canvas: Canvas) {
            this.canvas = canvas
        }

        fun draw() {
            if (isUseSelf) {
                canvas = holder.lockCanvas()
                canvas!!.drawColor(Color.WHITE)
            }
            //			canvas.drawText("123" +
//					"123", x, y, paint);
            val paint = Paint()
            paint.setTextSize(message!!.getMessageFont()!!.getSize().toFloat())
            val currentCanvas = canvas!!
            val g = GraphicsObject(currentCanvas, paint)

            message!!.createCustomUI(g, this@MessageBoxStringView.x, this@MessageBoxStringView.y)

            //			drawMessage(g, new LColor(100, 100, 100));
            AVGUtils.pause(30)

            message!!.update(30)

            if (isUseSelf) this@MessageBoxStringView.holder!!.unlockCanvasAndPost(currentCanvas)
        }

        var thread: Thread = Thread(object : Runnable {
            override fun run() {
                // TODO Auto-generated method stub
                while (true) {
                    draw()
                }
            }
        })

        init {
            this@MessageBoxStringView.holder = getHolder()
            this@MessageBoxStringView.holder!!.addCallback(this)

            this@MessageBoxStringView.holder!!.setFormat(PixelFormat.TRANSPARENT)
            setZOrderOnTop(true)
            setZOrderMediaOverlay(true)
        }

        override fun surfaceChanged(
            holder: SurfaceHolder, format: Int, width: Int,
            height: Int
        ) {
            // TODO Auto-generated method stub
        }

        override fun surfaceCreated(holder: SurfaceHolder) {
            // TODO Auto-generated method stub
            if (isUseSelf) thread.start()
            val lp = getLayoutParams()
            lp.width = width
            lp.height = height
            setLayoutParams(lp)
        }

        override fun surfaceDestroyed(holder: SurfaceHolder) {
            // TODO Auto-generated method stub
        }


        override fun onTouchEvent(event: MotionEvent?): Boolean {
            // TODO Auto-generated method stub

//			bb(event);

            return true
        }
    }
}
